snuba/datasets/discover.py

```diff
@@ -55,6 +61,67 @@
 logger = logging.getLogger(__name__)
 
 
+EVENT_CONDITION = FunctionCallMatch(
+    None,
+    Param("function", Or([StringMatch(op) for op in BINARY_OPERATORS])),
+    (
+        ColumnMatch(None, None, StringMatch("type")),
+        Param("event_type", AnyExpression()),
+    ),
+)
+
+
+def match_query_to_table(
+    query: Query, events_only_columns: ColumnSet, transactions_only_columns: ColumnSet
+) -> (str, bool, bool):
+    condition = query.get_condition_from_ast()
+    condition_columns = set()
+    event_types = []
+    if condition:
+        for cond in condition:
+            if Any(Column).match(cond):
+                condition_columns.add(cond)
+
+            result = EVENT_CONDITION.match(cond)
+            if not result:
+                continue
+
+            event_type_param = result.expression("event_type")
+            event_type = None
+            if isinstance(event_type_param, Column):
+                event_type = result.expression("event_type").column_name
+            else:
+                event_type = result.expression("event_type").value
+            if result:
+                if result.string("function") == ConditionFunctions.EQ:
+                    event_types.append(event_type)
+                elif result.string("function") == ConditionFunctions.NEQ:
+                    if event_type == "transaction":
+                        return EVENTS
+
+    if len(event_types) == 1 and event_types[0] == "transaction":
+        return TRANSACTIONS
+    elif len(event_types) >= 1:
+        return EVENTS
```

Comment:
I am not sure this condition is even needed.
Why not relying entirely on the columns present in the query?
- if there are common columns + errors columns -> errors
- if there are common columns + errors transactions -> transactions
- any other case events

if there are both errors and transactions columns -> impossible query.  Only then look into the condition, which is a less reliable condition anyway.

Reply:
The condition needs to be first, since that is defined by the user. We should respect what the user is telling us first, and then infer from columns.

Reply:
But if the user does not look for transactions specific columns, why do we need to rely on the transactions table ?
The events table contains both types of events, thus if you only look for common fields, you should obtain the same result when querying events or transactions. Am I right?
Which will work even better when we split errors and transactions in two tables since we will have a common merged table. https://clickhouse.tech/docs/en/engines/table-engines/special/merge/

Reply:
If the user has `event.type:transaction` in the filter of their query, then I think we should be using the transaction table regardless of their column selections. Both for the UX for our customers, and also for ourselves to have a guaranteed way to get data from one table or the other.

So you are right, but I think this way serves the end user better.